docs/proposals/004-scalability-benchmarking.md

```diff
@@ -0,0 +1,102 @@
+* This proposal does not intend to add testing that determines how a change impacts the scalability of Argo CD based on the benchmarks.
+* We do not intend to analyze cost implications of running different topologies and purely focus on scalability requirements from a technology perspective.
+
+## Proposal
+1.  Create new `argo-cd-benchmarking` repo under `argoproj-labs` and add authors of this proposal as maintainers.
+2. Create a set of key scalability factors to use as testing parameters. For example:
+    1. Number of Applications.
+    2. Number of resources managed by an Application.
+    3. Number of resources in a cluster.
+    4. The size of the resources in a cluster and managed by an Application.
+    5. Churn rate for resources in the cluster (how often resources change).
+    6. Number of clusters.
+    7. Number of repositories being monitored.
+    8. Size of the repositories.
+3. Determine the metrics that reflect limitations in scalability factors.
+    1. Application sync time for x number of apps
+    2.. Emptying the queues (app_reconciliation_queue, app_operation_processing_queue)
+4. Create automated testing procedures for Argo CD that take the key scalability factors as testing parameters.
+5. Test the default installation of Argo CD to determine the limit based on the key scalability factors.
+6. Create test scenarios that reflect the common topologies (Argo CD 1-1 with clusters, Argo CD 1-many with clusters).
+7. Determine the thresholds for the metrics identified earlier to capture when performance is being impacted.
+    1. Contribute back grafana thresholds and alerts for prometheus
+
+### Use cases
+Each use case will cover a specific topology with N permutations based on the key scalability factors. They will be measured using the metrics given in the proposal.
+
+We intend to focus on two key topologies: one Argo CD per cluster and one Argo CD for all remote clusters. All other variations are an extension of these.
+
+#### Topology 1: One Argo CD per cluster.
+The exact key scalability factors used in each permutations will be determined once we get to the testing.
+
+#### Topology 2: One Argo CD for all remote clusters.
+This will capture the impact of network throughput on performance.
+
+### Implementation Details/Notes/Constraints [optional]
+There is already some [tooling in the Argo CD repository](https://github.com/argoproj/argo-cd/pull/8037/files) for scalability testing. We plan to build on the existing effort and further push the boundaries of testing it.
+
+* Automatically set up supporting tooling for capturing metrics (grafana, prometheus)
+* Use a local Gittea in cluster to support many repos used in testing, and avoid performance variance by depending on the performance of a external git SaaS (ie GitHub)
+* Simulate the cluster and nodes using vcluster or kwok, in addition of using a cloud provider with a real cluster fleet.
```

Comment:
I'm curious - would the non-realworld-ness (sorry, not a term) of this setup impact the usefulness of the benchmark results ?

Reply:
Great question! The simulated clusters and nodes are intended to make the testing accessible, but ultimately we want the infrastructure to be easily changed to improve the benchmarks' "realworld-ness".

Once we have the benchmarking tooling, we can determine if the simulated components skew the results compared to the real world.

- [ ] I'll add your comment to that implementation detail point.

Reply:
Adding on top of @morey-tech already added, the simulated clusters will help for faster developing and iterating on the implementation of the tooling benchmark.